pkg/controller/node_availability.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ReconcileNodeAvailability struct {
+	client                client.Client
+	namespace             string
+	controllerSharedState *SharedState
+	azVolumeClient        azClientSet.Interface
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeAvailability{}
+
+func (r *ReconcileNodeAvailability) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		//Node is schedulable, proceed to attempt creation of replica attachment
+		r.controllerSharedState.priorityReplicaRequestsQueue.mut.Lock()
```

Comment:
Using a mutex to prevent re-entry into the queue processing loop is overkill and can lead to blocked goroutines that will end up doing nothing when they acquire the lock (i.e. first goruotine processes and clears the queue then releases the mutex. Unblocked goroutines now have nothing to do.)

You can use a the [atomic](https://pkg.go.dev/sync/atomic) package and a simple `int` flag in this case, e.g.

```go
// Set the flag to 1 and process the queue if it was previously 0
if atomic.SwapInt32(&r.controllerSharedState.processingReplicaRequestQueue, 1) == 0 {
  // Process the priority queue here

  // Set the flag back to 0 to allow other qoroutines to enter the processing loop
  atomic.StoreInt32(&r.controllerSharedState.processingReplicaRequestQueue, 0)
}
```